### PR TITLE
GEA 3248 audit tools - Multiples fixes

### PR DESCRIPTION
**Pull request by pangea-andrest**
- Fix search requests: SDK update search use datatime as param
- Fix hash verification: It didn't support fields with json
- Fix dump_after: was skipping first event and it doesn't not need to do that in all cases (border condition)

### CODE REVIEW

File: packages/pangea-sdk/README.md [hunk 1]
```diff
@@ -113,6 +113,8 @@ Verify that an event or a list of events has not been tampered with. Usage:
 
 ```
 usage: python -m pangea.verify_audit [-h] [--file PATH]
+or
+usage: poetry run python -m pangea.verify_audit [-h] [--file PATH]
```
Comment (dbenders) [inline packages/pangea-sdk/README.md:117]: not sure if we should add this to the README. poetry shouldn't be mandatory and we're not referencing it anywhere else

File: packages/pangea-sdk/README.md [hunk 2]
```diff
@@ -139,6 +141,8 @@ jsonl file (one json per line)
 
 ```
 usage: python -m pangea.dump_audit [-h] [--token TOKEN] [--domain DOMAIN] [--output OUTPUT] start end
+or
+usage: poetry run python -m pangea.dump_audit [-h] [--token TOKEN] [--domain DOMAIN] [--output OUTPUT] start end
```
Comment (dbenders) [inline packages/pangea-sdk/README.md:145]: same

File: packages/pangea-sdk/pangea/deep_verify.py [hunk 3]
```diff
@@ -77,7 +78,7 @@ def get_path_size(tree_size: int, leaf_index: int) -> int:
 def _tree_size_left(tree_size: int) -> int:
     """if the tree has size tree_size, return the size of the left child"""
     if tree_size <= 1:
-        return 0
+        return tree_size
```
Comment (dbenders) [inline packages/pangea-sdk/pangea/deep_verify.py:81]: 🤦

File: packages/pangea-sdk/pangea/deep_verify.py [hunk 4]
```diff
@@ -204,21 +203,24 @@ def deep_verify(audit: Audit, file: io.TextIOWrapper) -> Errors:
...
-                hot_out_of_order.add(hot_idx)
-
-        if hot_out_of_order:
-            errors["missing"] += len(hot_out_of_order)
+            hot_indexes.add(hot_idx)
+
+        hot_indexes_diff = set(range(len(events_by_idx))) - hot_indexes
+        if len(hot_indexes_diff) > 0:
+            errors["missing"] += len(hot_indexes_diff)
+            print(f"missing hot indexes: {hot_indexes_diff}")
```
Comment (dbenders) [inline packages/pangea-sdk/pangea/deep_verify.py:217]: I think this is too much information and too internal (we don't speak about hot/cold with the users)

File: packages/pangea-sdk/pangea/dump_audit.py [hunk 5]
```diff
@@ -66,44 +75,43 @@ def dump_before(audit: Audit, output: io.TextIOWrapper, start: datetime) -> int:
...
     cnt = 0
     if search_res.result.count > 0:
         leaf_index = search_res.result.events[0].leaf_index
-        for row in search_res.result.events[1:]:
-            if row.leaf_index != leaf_index:
-                break
-            dump_event(output, row, search_res)
-            cnt += 1
+        if leaf_index == last_leaf_index:
+            start = 1 if last_event_hash == search_res.result.events[0].hash else 0
```
Comment (dbenders) [inline packages/pangea-sdk/pangea/dump_audit.py:90]: good discovery!

File: packages/pangea-sdk/pangea/dump_audit.py [hunk 6]
```diff
@@ -66,44 +75,43 @@ def dump_before(audit: Audit, output: io.TextIOWrapper, start: datetime) -> int:
...
 
 
 def dump_page(
     audit: Audit, output: io.TextIOWrapper, start: datetime, end: datetime, first: bool = False
-) -> tuple[datetime, int]:
-
-    print("Dumping...", end="\r")
+) -> tuple[datetime, int, bool, str, int]:
+    PAGE_SIZE = 1000
+    print(start, end)
```
Comment (dbenders) [inline packages/pangea-sdk/pangea/dump_audit.py:104]: debug?

**Review by dbenders — APPROVED** [review]
minor comments, mainly for printing and doc. Code LGTM

File: packages/pangea-sdk/pangea/deep_verify.py [hunk 7]
```diff
@@ -102,10 +103,8 @@ def verify_hash(data: dict, data_hash: str) -> bool:
     """Verify the hash of an event"""
     succeeded = False
     try:
-        data_canon = audit_util.canonicalize_json(data)
-        computed_hash_dec = audit_util.hash_bytes(data_canon)
-        data_hash_dec = audit_util.decode_hash(data_hash)
-        if computed_hash_dec != data_hash_dec:
+        if not audit_util.verify_envelope_hash(EventEnvelope(**data), data_hash):
+            print("Hash failed: ", data)
```
Comment (ggallien13) [inline packages/pangea-sdk/pangea/deep_verify.py:107]: Can this be changed to "Hash verification failed: "? That would make the error clearer.

**Pull request merged**